### ISSUE TITLE
Decouple preprocessing, feature engineering, and training with reusable versioned datasets

### ISSUE THREAD

**Issue by linhsinp** [linhsinp/junyi-online-learning-prediction#7]
## Problem and outcome

Training currently orchestrates preprocessing, feature engineering, and model registration for every requested date window. Repeating a model experiment therefore repeats upstream computation and persistence. A reported test run covering approximately ten days took more than 15 hours without a clear failure diagnosis; the recently added logging provides a basis for measuring this, but the bottleneck has not yet been established.

Decouple the stages so daily ingestion maintains reusable processed data, feature jobs independently publish versioned datasets, and training consumes an existing feature dataset without invoking upstream stages.

**Primary acceptance criterion:** two training experiments against the same published feature snapshot perform zero preprocessing and zero feature recomputation, and both record the exact same input dataset identity.

## Current behavior

- `bootstrap/lake.py` converts a source CSV into monthly Parquet partitions and deletes the existing destination before rebuilding; it does not track daily arrivals.
- `workflows/training.py::training_pipeline` invokes all three stages. Its declared weekly trigger uses a default seven-day input window; the local Make target supplies explicit dates.
- Preprocessing appends to `processed_log` and writes run-specific Parquet artifacts. Feature engineering reads those artifacts rather than the processed table.
- Feature engineering appends relational fields to `feature_snapshot`, but the proficiency matrices required for training are stored separately as `.npy` artifacts.
- Training reads one run-specific feature artifact snapshot. Contracts and artifact keys use `training_run_id` across upstream stages.
- Database appends are not idempotent across overlapping runs or partial failures. Completed snapshots do not provide a general reuse/resume interface.
- The registry persists the current run's winning candidate and replaces `models/approved.json` without comparing that candidate with the existing approved model.

Relevant code: [workflow](https://github.com/linhsinp/junyi-online-learning-prediction/blob/main/src/junyi_predictor/workflows/training.py), [contracts](https://github.com/linhsinp/junyi-online-learning-prediction/blob/main/src/junyi_predictor/contracts.py), [feature builders](https://github.com/linhsinp/junyi-online-learning-prediction/blob/main/src/junyi_predictor/pipeline/feature_engineering.py), [lake materialization](https://github.com/linhsinp/junyi-online-learning-prediction/blob/main/src/junyi_predictor/bootstrap/lake.py), and [registry](https://github.com/linhsinp/junyi-online-learning-prediction/blob/main/src/junyi_predictor/registry.py). Update the existing architecture and runbook in `docs/implementation-plan.md`, `docs/current-system-design.md`, and `docs/logging.md` as this work lands.

## Target behavior

| Component | Trigger | Input selection | Durable output |
| --- | --- | --- | --- |
| Raw datalake | Daily file arrivals | Immutable CSV files with source identity/revision | Raw objects and ingestion metadata; optional reusable Parquet conversion |
| Preprocessing | Configurable nightly schedule or manual run | Unprocessed source revisions by default; explicit backfill/reprocessing supported | Versioned, deduplicated `processed` records in PostgreSQL |
| Feature engineering | Configurable weekly schedule or manual run | Newly processed events or an explicit historical range, with separately defined historical context | Complete feature records in PostgreSQL and published feature snapshot metadata |
| Training | Configurable weekly schedule or manual run | Select all eligible history for one feature definition by default, then split that dataset into training, validation, and test sets; optional event-date bounds or explicit snapshot restrict selection before splitting | Run metadata, metrics, scaler, model bundle, and registration in the artifact store |

**Meaning of “all data”:** select the full eligible feature dataset as the input to the chronological train-validation-test split. Fit the model and learned transformations on the training partition, use validation for candidate selection, and reserve the test partition for final evaluation. “All” describes the dataset selected before splitting; it does not mean fitting on validation or test rows. Apply any optional date filter before the split.

The logical table names above may be implemented by migrating the existing tables. Each component has an independent entrypoint, configuration, run identity, resource policy, and retry boundary. Keep an explicit end-to-end entrypoint for backfills and integration tests if useful; the standard training entrypoint must not call upstream stages.

### Current: one trigger executes all stages

```mermaid
flowchart TD
    CSV["Raw CSV"] --> BOOT["Separate bootstrap: rebuild Parquet"]
    BOOT --> LAKE[("Partitioned Parquet")]
    RUN["Weekly or manual combined workflow"] -.-> PRE["Preprocess selected window"]
    LAKE --> PRE
    PRE --> PS[("Run-specific preprocessing artifacts")]
    PS --> FE["Build features for the same run"]
    FE --> FS[("Run-specific feature artifacts and matrices")]
    FS --> TRAIN["Train, evaluate, and register"]
    TRAIN --> REG[("Model artifacts and approved pointer")]
    PRE --> PDB[("processed_log: appended output")]
    FE --> FDB[("feature_snapshot: partial feature output")]
```

The database outputs above are not the inputs consumed by the next stage. Each combined workflow invocation performs upstream computation again.

### Proposed: independent jobs consume published datasets

```mermaid
flowchart TD
    RAW[("Daily raw CSV in datalake")] --> PRE
    PSCHED["Nightly schedule or manual preprocessing"] -.-> PRE["Preprocess new source revisions<br/>Optional reusable Parquet conversion"]
    PRE --> PROCESSED[("Committed, versioned processed records<br/>PostgreSQL")]

    FSCHED["Weekly schedule or manual feature build"] -.-> FE["Build features<br/>New events or selected historical range"]
    PROCESSED --> FE
    HISTORY[("Historical context / versioned state")] --> FE
    FE --> FEATURES[("Published, complete feature datasets<br/>PostgreSQL + versioned mappings")]

    TSCHED["Weekly schedule or manual training"] -.-> SELECT["Pin one published feature snapshot<br/>Select all eligible data by default<br/>Apply optional date filters before splitting"]
    FEATURES --> SELECT
    SELECT --> SPLIT["Chronological train-validation-test split"]

    SPLIT --> TR["Training partition"]
    SPLIT --> VA["Validation partition"]
    SPLIT --> TE["Held-out test partition"]
    TR --> FIT["Fit scaler and candidate models"]
    FIT --> CHOOSE["Select candidate using validation metrics"]
    VA --> CHOOSE
    CHOOSE --> EVAL["Evaluate selected candidate on held-out test"]
    TE --> EVAL
    EVAL --> REGISTER["Register selected model<br/>Apply explicit approved-model promotion policy"]
    REGISTER --> ART[("Artifact store<br/>Model, scaler, metrics, dataset lineage,<br/>split metadata, registration and approved pointer")]
```

Solid arrows show data or artifact flow; dashed arrows show independent job triggers. Each stage consumes committed upstream publications. A training trigger starts dataset selection and splitting only; it never invokes preprocessing or feature engineering. Feature builds supply the historical context required by their definition, even when output is limited to newly arrived events. Registration records the selected model; changing the approved pointer follows the separately defined promotion policy.

## Design requirements

1. **Idempotent ingestion and preprocessing.** Track source identity/checksum, processing version, and completion status. Define stable event identity that preserves legitimate repeated attempts while deduplicating replayed input. Process once per source revision and preprocessing version; allow deliberate corrections and reprocessing. Commit data and completion metadata consistently so failed or concurrent jobs cannot expose partial success.
2. **Historical feature semantics.** Distinguish the output event window from the history used to calculate features. Existing cumulative accuracy and proficiency state resets for each supplied window. Establish correct historical replay first, then checkpoint state for incremental execution. Define deterministic event ordering, timestamp/timezone handling, late-arrival replay, and correction behavior. Features must use information available at prediction time, including an explicit policy for changing learner/content attributes.
3. **Versioned, complete feature datasets.** Separate feature-definition version, build/run identity, and published data revision. Persist all model inputs, target linkage, and stable feature mappings. Store proficiency values compactly instead of thousands of mostly-zero database columns. Experiments and overlapping builds must not silently duplicate training examples or overwrite datasets already referenced by a model.
4. **Reproducible training reads.** Resolve database inputs to a committed, reproducible snapshot at run start. Persist sufficient immutable data or revision references to reproduce it after later corrections. Record versions, event bounds, snapshot identity, row counts, feature ordering/mappings, split boundaries, code/configuration, and random seeds. A manifest alone is insufficient if referenced rows remain mutable.
5. **Evaluation and promotion.** Select all eligible examples for one compatible feature definition by default, then partition the selected dataset chronologically into disjoint training, validation, and test sets. Use validation for candidate selection and a separate test period for final reporting; fit learned transformations only on training data. Make comparison of full-history versus shorter-window input datasets possible, applying a train-validation-test split in either case. Register the run winner, and define an explicit promotion policy before replacing the existing approved model.
6. **Scheduling and readiness.** Flyte owns independent schedules; support manual execution, disabling schedules, and explicit schedule timezone configuration. Consumers read only published data and apply a documented freshness policy. Do not assume that offsetting weekly cron times guarantees upstream completion. Capture input publication IDs and distinguish empty inputs, stale inputs, failures, and successful no-op runs.

## Implementation milestones

- [ ] **M1 — Contracts and schema:** document event identity, versioning, snapshot publication, historical context, and late-data policy; add database migrations, uniqueness constraints, indexes, and independent stage contracts. Define retention of referenced revisions.
- [ ] **M2 — Independent preprocessing:** implement daily source discovery/manifest tracking, optional incremental Parquet conversion, idempotent processed-data writes, and manual preprocessing entrypoint. Preserve unrelated lake partitions.
- [ ] **M3 — Independent feature engineering:** read processed data from PostgreSQL; implement historical-range builds and complete compact feature persistence; publish immutable snapshot membership/mappings; add incremental state with equivalence tests against historical replay.
- [ ] **M4 — Independent training:** select published database features with optional date filters and all eligible history by default, then split the selected dataset into training, validation, and test partitions; persist reproducibility metadata; separate candidate evaluation, registration, and approved-model promotion. Add a training-only command and update existing callers.
- [ ] **M5 — Scheduling and recovery:** add configurable independent Flyte schedules, freshness checks, stage-specific retries/timeouts/resources, and recovery for partial writes and interrupted publication. Extend structured logging to independent run and dataset identities.
- [ ] **M6 — Migration, documentation, and validation:** document how existing run-scoped artifacts/tables are retained, migrated, or explicitly rebuilt; update Make targets and runbooks; run regression/acceptance checks and record staged performance measurements.

## Acceptance criteria

- [ ] **AC1:** Each stage runs manually and on its own configurable schedule. Invoking training never starts preprocessing or feature engineering.
- [ ] **AC2:** Replaying an unchanged raw file or overlapping preprocessing request adds no duplicate logical events; changed source/preprocessing versions follow the documented reprocessing policy. Legitimate repeated attempts remain distinct.
- [ ] **AC3:** Daily ingestion preserves previously committed raw/Parquet partitions and processed records. Failed or concurrent attempts leave no falsely completed publication or duplicate committed records.
- [ ] **AC4:** Feature jobs support newly arrived data and explicit historical windows. With the same context and feature version, incremental processing and full historical replay produce equivalent values and mappings within documented numeric tolerances.
- [ ] **AC5:** Late arrivals, corrections, equal timestamps, and changing dimensions follow tested deterministic rules without future-data leakage. Existing training snapshots remain reproducible after a rebuild.
- [ ] **AC6:** PostgreSQL contains the complete logical feature dataset needed for training. Training does not require an upstream training-run-specific `.npy` artifact or read raw/processed data to recompute features.
- [ ] **AC7:** By default, select all eligible rows from one feature definition as the dataset to split into chronological training, validation, and test partitions. Optional start-inclusive/end-exclusive date bounds filter that dataset before splitting. Exclude partial builds and duplicate experimental variants. The resolved input snapshot is pinned for the entire run; the three partitions are disjoint and together cover the selected eligible dataset.
- [ ] **AC8:** Two independent training runs can reuse one published snapshot while upstream entrypoints are unavailable. Both manifests identify the same data revision, mappings, and selection; neither changes upstream data.
- [ ] **AC9:** Record chronological split boundaries and row counts for all three partitions. Fit the model and learned transformations only on the training partition, select candidates on validation metrics, and use the held-out test partition only for final evaluation. Approved-model changes follow the documented policy, including when the existing model performs better.
- [ ] **AC10:** Failure recovery is independently scoped to the affected stage. Partial writes cannot become readable published datasets, and repeated registration/publication follows a defined idempotent outcome.
- [ ] **AC11:** Logs expose stage/run ID, input/output publication IDs, selected time range, row counts, duration, progress, and terminal status without raw learner records or credentials. Freshness/no-data decisions are visible.
- [ ] **AC12:** Relevant regression tests, the full `uv run pytest` suite, and lint checks pass. One-, three-, and ten-day benchmark results record hardware/resources, input sizes, per-stage durations, peak memory, and reused versus recomputed work; remaining bottlenecks are documented without assuming a speedup target before measurement.

## Required testing scenarios

| Scenario | Expected evidence |
| --- | --- |
| Duplicate file, overlapping date range, legitimate repeated attempt | Stable event counts after replay; distinct genuine attempts preserved. |
| Changed file contents or preprocessing version | Deliberate revision/reprocessing; prior referenced data remains reproducible. |
| Crash after writing rows but before publishing; retry; concurrent overlapping jobs | No partial published dataset, duplicate rows, or incorrect completion marker. Test against PostgreSQL, not only mocks. |
| Incremental daily/weekly batches versus one full historical build | Matching feature values, row identities, and column mappings with identical context. |
| Late event, corrected outcome, new user/concept, tied timestamps, changing dimensions | Deterministic ordering, appropriate replay/versioning, stable schemas, and no future-information leakage. |
| Historical experiment with overlapping dates or a new feature definition | Isolated outputs; no duplicate examples or mutation of another published snapshot. |
| Default full-history selection followed by train-validation-test split; explicit date bounds; empty or stale data | Without date bounds, the union of the three disjoint partitions equals all eligible rows in the pinned snapshot. With bounds, it equals the eligible rows in that interval. Verify filters run before splitting, recorded partition counts/boundaries, and documented no-data/freshness outcomes. |
| Two training runs against one publication with upstream calls configured to fail if invoked | Both complete using stored features; zero upstream calls and identical dataset lineage. |
| Feature publication/correction occurs while training loads inputs | Training remains on its pinned revision with stable counts and mappings. |
| Chronological evaluation and a weaker candidate than the approved model | Model/scaler fitting receives only training rows; validation rows are used for selection and test rows only for final evaluation. No split/scaler leakage; promotion policy is enforced. |
| Failure during artifact persistence or approved-pointer update; retry | Recoverable registration with no approved pointer to an incomplete bundle. |
| Independent schedules/manual runs and full optional composition | Correct entrypoints and publication readiness; schedule timezone/configuration covered without running production jobs. |
| Existing-data migration/rebuild and representative performance runs | Documented transition with preserved lineage where promised; measured runtime/memory and confirmed upstream reuse. |

## Scope boundaries

This issue covers stage decoupling, reusable data contracts, correctness, recovery, and the validation needed to operate them. It does not promise to eliminate the reported 15-hour runtime, introduce distributed training/GPU infrastructure, or prove that selecting full history before the train-validation-test split gives the best model. Use the resulting measurements to open focused performance follow-ups. A new feature-store service is not required; PostgreSQL remains the requested source for processed and feature data.
